### PR TITLE
Account activation and password reset

### PR DESCRIPTION
**Pull request by agalkiewicz**

### CODE REVIEW

File: app/controllers/password_resets_controller.rb [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+  before_action :get_user,         only: [:edit, :update]
+  before_action :valid_user,       only: [:edit, :update]
+  before_action :check_expiration, only: [:edit, :update]
+
+  def new
+  end
+
+  def create 
+    @user = User.find_by(email: params[:password_reset][:email].downcase)
+    if @user
```
Comment (mmarcinkowski) [inline app/controllers/password_resets_controller.rb]: WDYF about `if @user = User.find_by(email: params[:password_reset][:email].downcase)`

File: app/controllers/password_resets_controller.rb [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+      @user.send_password_reset_email
+      flash[:info] = "Email sent with password reset instructions"
+      redirect_to root_url
+    else
+      flash.now[:danger] = "Email address not found"
+      render 'new'
+    end
+  end
+
+  def edit
```
Comment (mmarcinkowski) [inline app/controllers/password_resets_controller.rb]: def edit; end

File: app/controllers/password_resets_controller.rb [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    end
+  end
+
+  private
+
+    def user_params
+      params.require(:user).permit(:password, :password_confirmation)
+    end
+
+    # Before filters
```
Comment (mmarcinkowski) [inline app/controllers/password_resets_controller.rb]: i think you can remove these comments. Code is self explained here

File: app/models/user.rb [hunk 4]
```diff
@@ -26,13 +27,55 @@ def remember
...
-  def authenticated?(remember_token)
-    return false if remember_digest.nil?
-    BCrypt::Password.new(remember_digest).is_password?(remember_token)
+  def authenticated?(attribute, token)
+    digest = send("#{attribute}_digest")
+    return false if digest.nil?
+    BCrypt::Password.new(digest).is_password?(token)
   end
 
   # Forgets a user.
```
Comment (mmarcinkowski) [inline app/models/user.rb]: This same as above

File: app/views/password_resets/edit.html.erb [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+    <%= form_for(@user, url: password_reset_path(params[:id])) do |f| %>
+      <%= render 'shared/error_messages' %>
+
+      <%= hidden_field_tag :email, @user.email %>
+
+      <%= f.label :password %>
+      <%= f.password_field :password, class: 'form-control' %>
+
+      <%= f.label :password_confirmation, "Confirmation" %>
+      <%= f.password_field :password_confirmation, class: 'form-control' %>
```
Comment (mmarcinkowski) [inline app/views/password_resets/edit.html.erb]: please use double quotation marks

File: app/views/password_resets/new.html.erb [hunk 6]
```diff
@@ -0,0 +1,13 @@
+<% provide(:title, "Forgot password") %>
+<h1>Forgot password</h1>
+
+<div class="row">
+  <div class="col-md-6 col-md-offset-3">
+    <%= form_for(:password_reset, url: password_resets_path) do |f| %>
+      <%= f.label :email %>
+      <%= f.email_field :email, class: 'form-control' %>
```
Comment (mmarcinkowski) [inline app/views/password_resets/new.html.erb]: this same as above

File: app/views/sessions/new.html.erb [hunk 7]
```diff
@@ -9,6 +9,7 @@
       <%= f.email_field :email, class: 'form-control' %>
 
       <%= f.label :password %>
+      <%= link_to "(forgot password)", new_password_reset_path %>
       <%= f.password_field :password, class: 'form-control' %>
```
Comment (mmarcinkowski) [inline app/views/sessions/new.html.erb]: this same as above

File: config/routes.rb [hunk 8]
```diff
@@ -8,4 +8,6 @@
   post   'login'   => 'sessions#create'
   delete 'logout'  => 'sessions#destroy'
   resources :users
+  resources :account_activations, only: [:edit]
+  resources :password_resets,     only: [:new, :create, :edit, :update]
```
Comment (mmarcinkowski) [inline config/routes.rb]: `%i(new create edit update)`

File: config/routes.rb [hunk 9]
```diff
@@ -8,4 +8,6 @@
   post   'login'   => 'sessions#create'
   delete 'logout'  => 'sessions#destroy'
   resources :users
+  resources :account_activations, only: [:edit]
```
Comment (mmarcinkowski) [inline config/routes.rb]: %i(edit)

File: db/migrate/20151101101142_add_activation_to_users.rb [hunk 10]
```diff
@@ -0,0 +1,7 @@
+class AddActivationToUsers < ActiveRecord::Migration
+  def change
+    add_column :users, :activation_digest, :string
+    add_column :users, :activated, :boolean, default: false
```
Comment (mmarcinkowski) [inline db/migrate/20151101101142_add_activation_to_users.rb]: maybe add `null: false` as well

File: test/integration/password_resets_test.rb [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+class PasswordResetsTest < ActionDispatch::IntegrationTest
+  
+  def setup 
+    ActionMailer::Base.deliveries.clear
+    @user = users(:michael)
+  end
+
+  test "password resets" do
+    get new_password_reset_path
+    assert_template 'password_resets/new'
```
Comment (mmarcinkowski) [inline test/integration/password_resets_test.rb]: quotation marks

File: test/integration/password_resets_test.rb [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+    user.toggle!(:activated)
+    # Right email, wrong token
+    get edit_password_reset_path('wrong token', email: user.email)
+    assert_redirected_to root_url
+    # Right email, right token
+    get edit_password_reset_path(user.reset_token, email: user.email)
+    assert_template 'password_resets/edit'
+    assert_select "input[name=email][type=hidden][value=?]", user.email
+    # Invalid password & confirmation
+    patch password_reset_path(user.reset_token), email: user.email, user: { password: "foobaz", password_confirmation: "barquux" }
```
Comment (mmarcinkowski) [inline test/integration/password_resets_test.rb]: line is too long

File: test/integration/password_resets_test.rb [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+    assert_select 'div#error_explanation'
+    # Empty password
+    patch password_reset_path(user.reset_token), email: user.email, user: { password: "", password_confirmation: "" }
+    assert_select 'div#error_explanation'
+    # Valid password & confirmation
+    patch password_reset_path(user.reset_token), email: user.email, user: { password: "foobaz", password_confirmation: "foobaz" }
```
Comment (mmarcinkowski) [inline test/integration/password_resets_test.rb]: this same as above

File: test/integration/users_signup_test.rb [hunk 14]
```diff
@@ -9,14 +13,32 @@ class UsersSignupTest < ActionDispatch::IntegrationTest
                                password_confirmation: "bar" }
     end
     assert_template 'users/new'
+    assert_select 'div#error_explanation'
```
Comment (mmarcinkowski) [inline test/integration/users_signup_test.rb]: quotation marks

File: test/mailers/previews/user_mailer_preview.rb [hunk 15]
```diff
@@ -0,0 +1,17 @@
+# Preview all emails at http://localhost:3000/rails/mailers/user_mailer
+class UserMailerPreview < ActionMailer::Preview
+
+  # Preview this email at http://localhost:3000/rails/mailers/user_mailer/account_activation
```
Comment (mmarcinkowski) [inline test/mailers/previews/user_mailer_preview.rb]: remove comments

File: app/controllers/password_resets_controller.rb (hunk 2 above)
Comment (agalkiewicz) [inline app/controllers/password_resets_controller.rb]: Changed.

File: app/controllers/password_resets_controller.rb (hunk 1 above)
Comment (agalkiewicz) [inline app/controllers/password_resets_controller.rb]: Changed.

File: app/controllers/password_resets_controller.rb (hunk 3 above)
Comment (agalkiewicz) [inline app/controllers/password_resets_controller.rb]: Deleted.

File: app/models/user.rb (hunk 4 above)
Comment (agalkiewicz) [inline app/models/user.rb]: Deleted.

File: config/routes.rb (hunk 9 above)
Comment (mmarcinkowski) [inline config/routes.rb]: u can even ommit these brackets, i mean `only: :edit`

File: app/views/password_resets/edit.html.erb (hunk 5 above)
Comment (agalkiewicz) [inline app/views/password_resets/edit.html.erb]: Done.

File: app/views/password_resets/new.html.erb (hunk 6 above)
Comment (agalkiewicz) [inline app/views/password_resets/new.html.erb]: Done.

File: app/views/sessions/new.html.erb (hunk 7 above)
Comment (agalkiewicz) [inline app/views/sessions/new.html.erb]: Done.

File: config/routes.rb (hunk 8 above)
Comment (agalkiewicz) [inline config/routes.rb]: Changed.

File: config/routes.rb (hunk 9 above)
Comment (agalkiewicz) [inline config/routes.rb]: Changed.

File: test/integration/password_resets_test.rb (hunk 11 above)
Comment (agalkiewicz) [inline test/integration/password_resets_test.rb]: Changed.

Comment (agalkiewicz) [inline test/integration/password_resets_test.rb]: Changed.

File: test/integration/password_resets_test.rb (hunk 12 above)
Comment (agalkiewicz) [inline test/integration/password_resets_test.rb]: Changed.

File: test/integration/password_resets_test.rb (hunk 13 above)
Comment (agalkiewicz) [inline test/integration/password_resets_test.rb]: Changed.

File: test/integration/users_signup_test.rb (hunk 14 above)
Comment (agalkiewicz) [inline test/integration/users_signup_test.rb]: Changed.

File: test/mailers/previews/user_mailer_preview.rb (hunk 15 above)
Comment (agalkiewicz) [inline test/mailers/previews/user_mailer_preview.rb]: I think these comments are useful. I'm not going to delete them. :)

File: db/migrate/20151101101142_add_activation_to_users.rb (hunk 10 above)
Comment (agalkiewicz) [inline db/migrate/20151101101142_add_activation_to_users.rb]: Added.

File: test/mailers/previews/user_mailer_preview.rb (hunk 15 above)
Comment (mmarcinkowski) [inline test/mailers/previews/user_mailer_preview.rb]: ech ech ech :sob:

**Pull request merged**